Added ShouldBeJsonSerializable

rikrak opened the pull request:
Added ShouldBeJsonSerializable (see #3 )
Note: I've added the extension to the root namespace (__FluentAssertions__) to aid discovery.

This is based on a previous PR[1] by @borismod that was never merged

[1] https://github.com/fluentassertions/fluentassertions.json/commit/54432fdebc39e8ac52f5c9a0a30c08bdab01bbcf

dennisdoomen commented on Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs:
```diff
@@ -0,0 +1,97 @@
+            {
+                Execute.Assertion
+                    .BecauseOf(because, becauseArgs)
+                    .FailWith("Expected {0} to be JSON serializable{reason}, but serialization failed with:\r\n\r\n{1}",
```
I recommend rewriting the failure as

`FailWith("Expected {object:context} to be JSON serializable{reason}, but serializing {0} failed with {1}", assertions.Subject, exc)`

This will make sure that the name of the variable on which you call `Should()` will be included in the message. But you need to add `[CustomAssertion]` to all your new methods. And by passing in the entire exception object, you let FA figure out how to format that the best way.

rikrak replied:
Thanks!  I've updated the PR as suggested

dennisdoomen commented on Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs:
```diff
@@ -0,0 +1,95 @@
+        {
+            var serializedObject = JsonConvert.SerializeObject(subject);
+            var cloneUsingJsonSerializer = JsonConvert.DeserializeObject(serializedObject, subject.GetType());
+            return cloneUsingJsonSerializer; ;
```
🙃 Nitpick: unnecessary semi-colon

dennisdoomen commented on Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs:
```diff
@@ -0,0 +1,95 @@
+                    .IncludingFields()
+                    .IncludingProperties();
+
+                var typedSubject = (T)assertions.Subject;
```
🤔 What if the subject is not of type `T`?
🤔 What if the subject is `null`?

rikrak replied:
With respect to what to do if the subject is `null`: Both `BeBinarySerializable()` and `BeDataContractSerializable()` fail the assertion with what looks like a `NullReferenceException`. I could do the same, but I was thinking of something more like this:
```
    Execute.Assertion.ForCondition(assertions.Subject != null)
        .BecauseOf(because, becauseArgs)
        .FailWith("Expected {object:context} to be JSON serializable{reason}, but the value is null.  Please provide a value for the assertion.");
```
Not quite sure on the fail message, but it gets the point across I think

dennisdoomen replied:
Yeah, that's sloppy old code. The main library will do exactly what you suggested.

rikrak replied:
Done :-)

dennisdoomen commented on Tests/FluentAssertions.Json.Specs/ShouldBeJsonSerializableTests.cs:
```diff
@@ -0,0 +1,89 @@
+            Action act = () => target.Should().BeJsonSerializable();
+
+            // assert
+            act.Should().Throw<Xunit.Sdk.XunitException>();
```
🤔 I think we should also verify a part of the message to make sure the test fails for the right reasons. Same for the other tests where you expect an exception.

dennisdoomen commented on Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs:
```diff
@@ -0,0 +1,95 @@
+
+// added to the root namespace to aid discovery
+// ReSharper disable CheckNamespace
+namespace FluentAssertions
```
❌ Please use the same namespace as the rest of the code base.

rikrak replied:
Fixed in https://github.com/fluentassertions/fluentassertions.json/pull/52/commits/c4ee603f6e7a2ec108af29611af0fd1f715c6715

dennisdoomen commented on Tests/FluentAssertions.Json.Specs/ShouldBeJsonSerializableTests.cs:
```diff
@@ -0,0 +1,89 @@
+        }
+
+        [Fact]
+        public void Class_is_simple_poco_should_be_serializable()
```
🤔 I don't quite get this naming convention in all of these test cases. For this one, did you intend to say "Class as simple poco should be serializable"? Maybe "Simple_poco_should_be_serializable" or "When_class_is_simple_poco_it_should_be_serializable"?

dennisdoomen reviewed (changes requested):
Apologies, but I noticed I wasn't finished with the review.

rikrak commented:
I'm not too familiar with the etiquette of github - Should I mark the review comments as _Resolved_ after a commit, or is that up to the reviewer?  Either way I've just (hopefully!) addressed the review comments with the latest commits - let me know if they're OK, or if there are any other bits to look at :-)

dennisdoomen commented:
Thank you for your contribution.

dennisdoomen commented:
Crap. Completely forgot to ask @jnyrup to also take a peek...

jnyrup commented on Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs:
```diff
@@ -0,0 +1,103 @@
+    public static class ObjectAssertionsExtensions
+    {
+        /// <summary>
+        /// Asserts that an object can be serialized and deserialized using the JSON serializer and that it stills retains
```
"stills retains" -> "still retains"

jnyrup commented on Tests/FluentAssertions.Json.Specs/ShouldBeJsonSerializableTests.cs:
```diff
@@ -0,0 +1,136 @@
+            // assert
+            act.Should()
+                .Throw<Xunit.Sdk.XunitException>(because:"This is consistent with BeBinarySerializable() and BeDataContractSerializable()")
+                .Which.Message
+                    .Should().Contain("value is null")
+                        .And.Contain("Please provide a value for the assertion");
```
Tip, it should be possible to shorten this to:
```cs
.WithMessage("*value is null*Please provide a value for the assertion*");
```

jnyrup commented on Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs:
```diff
@@ -0,0 +1,103 @@
+            return new AndConstraint<ObjectAssertions>(assertions);
+        }
+
+        private static object CreateCloneUsingJsonSerializer(object subject)
+        {
+            var serializedObject = JsonConvert.SerializeObject(subject);
+            var cloneUsingJsonSerializer = JsonConvert.DeserializeObject(serializedObject, subject.GetType());
+            return cloneUsingJsonSerializer;
+        }
```
Does it make any difference if we use `subject.GetType()` or `typeof(T)`?
E.g. for inheritance and interfaces.

rikrak replied:
Looks like using typeof(T) causes issues with the non-generic overload of BeJsonSerialisable as `T` is `object` which causes problems with the equivalency of the deserialised object.  Essentially a dynamic object is deserialised that doesn't play well with the IsEquivalentTo test.  In addition the deserialised dynamic object has only string properties rather than, for example, Guids or DateTimes that the original instance had.

jnyrup reviewed (commented):
I was thinking out giving this a view later today.

Aside from the single question about inheritance, it looks good to me.
